### PR TITLE
Added margin property to FormField

### PR DESCRIPTION
**Pull request by joybhallaa**
#### Any background context you want to provide?
Issue #3109 

I have tried to address the issue raise by @ericsoderberghp in #3109 , I have added right,top,left margin variables and then they will be passed in the components.
#### Do the grommet docs need to be updated?
No
#### Should this PR be mentioned in the release notes?
Not sure
#### Is this change backwards compatible or is it a breaking change?
Not sure

### CODE REVIEW

File: src/js/components/FormField/FormField.js [hunk 1]
```diff
@@ -171,14 +171,29 @@ class FormFieldContent extends Component {
...
         let marginBottom = '-1px';
+        let marginTop = '0px';
+        let marginRight = '0px';
+        let marginLeft = '0px';
         if (border.size) {
           marginBottom = `-${parseMetricToNum(
             theme.global.borderSize[border.size],
           )}px`;
+          marginTop = `-${parseMetricToNum(
+            theme.global.borderSize[border.size],
```
Comment (oorestisime) [inline src/js/components/FormField/FormField.js]: maybe we can avoid doing this calculation 4 times

**Comment by joybhallaa** [comment]
Sure, will work on it in the upcoming days.

File: src/js/components/FormField/FormField.js (hunk 1 above)
Comment (joybhallaa) [reply src/js/components/FormField/FormField.js]: Okay, will fix this.

Comment (joybhallaa) [reply src/js/components/FormField/FormField.js]: @oorestisime something in line of 
```
static defaultProps = {
 margin: {
  top: 'top',
  bottom: 'bottom',
  right: 'right',
  left: 'left',
 }
}
```

will work ?

**Comment by joybhallaa** [comment]
Hey @oorestisime , so as you said I made margin a prop, and extended it's utility in *base.js*, by doing 

``` 
margin : {
top: 'small',
bottom: 'small,
left: 'small,
right: 'small',
}
```

I also added *margin* to the props,  and then rendering it in the FormFieldBox component, I have also added a test, and updated the snapshot. 
Please see the changes from the last commit and let me know if it's right or not, and give feedback.

File: src/js/components/Form/__tests__/__snapshots__/Form-test.js.snap [hunk 2]
```diff
@@ -277,7 +270,6 @@ exports[`Form with field 1`] = `
   box-sizing: border-box;
   outline: none;
   max-width: 100%;
-  margin-bottom: 12px;
```
Comment (ShimiSun) [inline src/js/components/Form/__tests__/__snapshots__/Form-test.js.snap]: do we know why this snapshot changed? it looks like a potential ui regression.

File: src/js/components/FormField/__tests__/FormField-test.js [hunk 3]
```diff
@@ -64,6 +64,16 @@ test('renders htmlFor', () => {
   expect(tree).toMatchSnapshot();
 });
 
+test('renders margin', () => {
```
Comment (ShimiSun) [inline src/js/components/FormField/__tests__/FormField-test.js]: Thanks for adding a test!

File: src/js/components/Form/__tests__/__snapshots__/Form-test.js.snap (hunk 2 above)
Comment (joybhallaa) [reply src/js/components/Form/__tests__/__snapshots__/Form-test.js.snap]: As far as I know, when I added it **margin** as a prop, and tried a test, it showed this change. This was before the update of the snapshot, I can try to find what's causing this change. 
Thanks for the review and feedback @ShimiSun!

Comment (joybhallaa) [reply src/js/components/Form/__tests__/__snapshots__/Form-test.js.snap]: @ShimiSun , if I remove the *margin* from the **props** object and pass **{...formField.margin}** instead, *margin-top,margin-bottom,margin-right,margin-left* are being added to my snapshot when running **yarn test**.

Screenshot from 2019-05-31 14-33-27

**Comment by joybhallaa** [comment]
@ShimiSun @oorestisime I'll really like your input on the new commit.

**Comment by oorestisime** [comment]
Hey! I am currently without PC until thursday! Will be the first thing i ll
look! Thanks for working on this!

File: src/js/components/FormField/__tests__/FormField-test.js (hunk 3 above)
Comment (joybhallaa) [reply src/js/components/FormField/__tests__/FormField-test.js]: Please review the latest commit.

**Comment by oorestisime** [comment]
Thanks once again for working on this. will take a closer look tonight. A few thoughts:

* right now you added values in the theme. that means all the formfields will have by default a margin at top left and right which is a breaking change (not sure intended)
* might have missed this but are we able to set margin via the props? it seems it is only available from the theme no?
* would like a bit more tests with different margin values to be sure we don't break anything now or later.

if you want we can pair on this when you have some time, on slack or anywhere you like!

File: src/js/components/Form/__tests__/__snapshots__/Form-test.js.snap [hunk 4]
```diff
@@ -76,7 +76,10 @@ exports[`Form update 1`] = `
   box-sizing: border-box;
   outline: none;
   max-width: 100%;
+  margin-top: 12px;
```
Comment (ericsoderberghp) [inline src/js/components/Form/__tests__/__snapshots__/Form-test.js.snap]: This looks like a breaking change.

File: src/js/components/FormField/FormField.js [hunk 5]
```diff
@@ -169,16 +168,8 @@ class FormFieldContent extends Component {
         border.position === 'outer' &&
         (border.side === 'all' || border.side === 'horizontal' || !border.side);
       if (abut) {
-        // marginBottom is set to overlap adjacent fields
```
Comment (ericsoderberghp) [inline src/js/components/FormField/FormField.js]: This is also a breaking change. We need to handle the case where the border is on all sides, and the focus indicator is visible appropriately, when there is no margin between the fields.

File: src/js/themes/base.js [hunk 6]
```diff
@@ -524,7 +524,12 @@ export const generate = (baseSpacing = 24, scale = 6) => {
       label: {
         margin: { vertical: 'xsmall', horizontal: 'small' },
       },
-      margin: { bottom: 'small' },
+      margin: {
```
Comment (ericsoderberghp) [inline src/js/themes/base.js]: This is a breaking change. Why do we need to change the default margin in the base theme?

**Comment by ericsoderberghp** [comment]
I agree with @oorestisime that we should support a margin in the theme as well as in the props.

**Comment by joybhallaa** [comment]
> Thanks once again for working on this. will take a closer look tonight. A few thoughts:
> 
> * right now you added values in the theme. that means all the formfields will have by default a margin at top left and right which is a breaking change (not sure intended)
> * might have missed this but are we able to set margin via the props? it seems it is only available from the theme no?
> * would like a bit more tests with different margin values to be sure we don't break anything now or later.
> 
> if you want we can pair on this when you have some time, on slack or anywhere you like!

Sure @oorestisime , I would love to pair with you on this, I will go through the code once again. Let me know when you are free? Is it okay to send you a personal message on slack ?

**Comment by joybhallaa** [comment]
@ericsoderberghp  thanks for the review, I will work on this with @oorestisime , and will try to get this implemented as soon as possible !

**Comment by oorestisime** [comment]
sure thing! ping me there!

File: src/js/components/FormField/FormField.js (hunk 5 above)
Comment (oorestisime) [reply src/js/components/FormField/FormField.js]: so this change needs to get reverted @joybhallaa

File: src/js/components/FormField/FormField.js [hunk 7]
```diff
@@ -193,7 +185,7 @@ class FormFieldContent extends Component {
             ? { ...border, color: borderColor }
             : undefined
         }
-        margin={abut ? undefined : { ...formField.margin }}
+        margin={abut ? margin : { ...formField.margin }}
```
Comment (oorestisime) [inline src/js/components/FormField/FormField.js]: this doesn't sound right to me. i think we should apply margin if not abut. something like `margin={abut ? undefined : margin || { ...formField.margin }}` so that we apply the margin property if user wants, otherwise we keep formField.margin for backwards compatibility.

**Comment by joybhallaa** [comment]
I will not be able to work on this feature, my laptop has broken down unexpectedly, and I don't know when it will get fixed.

**Comment by buoyantair** [comment]
@joybhallaa Ah, sorry that happened, I know exactly how it feels. It's fine if you aren't able to do it now, someone else can pick it up :D Magic ✨ of open source ^^

**Comment by oorestisime** [comment]
hey @joybhallaa how is your pc situation going? we can wait a bit more for this if you wish to finish it :) take care!

**Comment by joybhallaa** [comment]
I'm sorry I can't help you guys anymore, should I close this PR?

**Comment by oorestisime** [comment]
No I will take over! thanks!

**Comment by oorestisime** [comment]
@ShimiSun @ericsoderberghp 
SO recap on what i did

- rebased to current master
- refactor margin prop into separate function in generic props so i could use it in formfield
- updated TS
- added a margin prop that precedes all options. If user passes a margin prop no matter what we will render with it.
- if abut and no margin prop i kept current state of things to avoid breaking changes
- if not abut and no margin prop we use theme value.
- added a few tests around this. 

I need a few more eyes on this, some edge cases i had to cover :)

**Comment by ShimiSun** [comment]
Thank you for collaborating on this @joybhallaa & @oorestisime! another great contribution 💟
